File: clients/python/model/building_properties.py
from model.resource import Resource
from stream_wrapper import StreamWrapper
from typing import Dict
from typing import Optional

class BuildingProperties:
    """TODO - Document"""

    __slots__ = ("build_resources","max_health","max_workers","work_resources","produce_worker","produce_resource","produce_amount","produce_score","harvest","work_amount",)

    build_resources: Dict[Resource, int]
    max_health: int
    max_workers: int
    work_resources: Dict[Resource, int]
    produce_worker: bool
    produce_resource: Optional[Resource]
    produce_amount: int
    produce_score: int
    harvest: bool
    work_amount: int

    def __init__(self, build_resources: Dict[Resource, int], max_health: int, max_workers: int, work_resources: Dict[Resource, int], produce_worker: bool, produce_resource: Optional[Resource], produce_amount: int, produce_score: int, harvest: bool, work_amount: int):
        self.build_resources = build_resources
        """TODO - Document"""
        self.max_health = max_health
        """TODO - Document"""
        self.max_workers = max_workers
        """TODO - Document"""
        self.work_resources = work_resources
        """TODO - Document"""
        self.produce_worker = produce_worker
        """TODO - Document"""
        self.produce_resource = produce_resource
        """TODO - Document"""
        self.produce_amount = produce_amount
        """TODO - Document"""
        self.produce_score = produce_score
        """TODO - Document"""
        self.harvest = harvest
        """TODO - Document"""
        self.work_amount = work_amount
        """TODO - Document"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "BuildingProperties":
        """Read BuildingProperties from input stream
        """
        build_resources = {}
        for _ in range(stream.read_int()):
            build_resources_key = Resource(stream.read_int())
            build_resources_value = stream.read_int()
            build_resources[build_resources_key] = build_resources_value
        max_health = stream.read_int()
        max_workers = stream.read_int()
        work_resources = {}
        for _ in range(stream.read_int()):
            work_resources_key = Resource(stream.read_int())
            work_resources_value = stream.read_int()
            work_resources[work_resources_key] = work_resources_value
        produce_worker = stream.read_bool()
        if stream.read_bool():
            produce_resource = Resource(stream.read_int())
        else:
            produce_resource = None
        produce_amount = stream.read_int()
        produce_score = stream.read_int()
        harvest = stream.read_bool()
        work_amount = stream.read_int()
        return BuildingProperties(build_resources, max_health, max_workers, work_resources, produce_worker, produce_resource, produce_amount, produce_score, harvest, work_amount)
    
    def write_to(self, stream: StreamWrapper):
        """Write BuildingProperties to output stream
        """
        stream.write_int(len(self.build_resources))
        for key, value in self.build_resources.items():
            stream.write_int(key)
            stream.write_int(value)
        stream.write_int(self.max_health)
        stream.write_int(self.max_workers)
        stream.write_int(len(self.work_resources))
        for key, value in self.work_resources.items():
            stream.write_int(key)
            stream.write_int(value)
        stream.write_bool(self.produce_worker)
        if self.produce_resource is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            stream.write_int(self.produce_resource)
        stream.write_int(self.produce_amount)
        stream.write_int(self.produce_score)
        stream.write_bool(self.harvest)
        stream.write_int(self.work_amount)
    
    def __repr__(self):
        return "BuildingProperties(" + \
            repr(self.build_resources) + \
            ", " + \
            repr(self.max_health) + \
            ", " + \
            repr(self.max_workers) + \
            ", " + \
            repr(self.work_resources) + \
            ", " + \
            repr(self.produce_worker) + \
            ", " + \
            repr(self.produce_resource) + \
            ", " + \
            repr(self.produce_amount) + \
            ", " + \
            repr(self.produce_score) + \
            ", " + \
            repr(self.harvest) + \
            ", " + \
            repr(self.work_amount) + \
            ")"

File: clients/cpp/model/BuildingProperties.cpp
#include "BuildingProperties.hpp"

namespace model {

BuildingProperties::BuildingProperties(std::unordered_map<model::Resource, int> buildResources, int maxHealth, int maxWorkers, std::unordered_map<model::Resource, int> workResources, bool produceWorker, std::optional<model::Resource> produceResource, int produceAmount, int produceScore, bool harvest, int workAmount) : buildResources(buildResources), maxHealth(maxHealth), maxWorkers(maxWorkers), workResources(workResources), produceWorker(produceWorker), produceResource(produceResource), produceAmount(produceAmount), produceScore(produceScore), harvest(harvest), workAmount(workAmount) { }

// Read BuildingProperties from input stream
BuildingProperties BuildingProperties::readFrom(InputStream& stream) {
    size_t buildResourcesSize = stream.readInt();
    std::unordered_map<model::Resource, int> buildResources = std::unordered_map<model::Resource, int>();
    buildResources.reserve(buildResourcesSize);
    for (size_t buildResourcesIndex = 0; buildResourcesIndex < buildResourcesSize; buildResourcesIndex++) {
        model::Resource buildResourcesKey = readResource(stream);
        int buildResourcesValue = stream.readInt();
        buildResources.emplace(std::make_pair(buildResourcesKey, buildResourcesValue));
    }
    int maxHealth = stream.readInt();
    int maxWorkers = stream.readInt();
    size_t workResourcesSize = stream.readInt();
    std::unordered_map<model::Resource, int> workResources = std::unordered_map<model::Resource, int>();
    workResources.reserve(workResourcesSize);
    for (size_t workResourcesIndex = 0; workResourcesIndex < workResourcesSize; workResourcesIndex++) {
        model::Resource workResourcesKey = readResource(stream);
        int workResourcesValue = stream.readInt();
        workResources.emplace(std::make_pair(workResourcesKey, workResourcesValue));
    }
    bool produceWorker = stream.readBool();
    std::optional<model::Resource> produceResource = std::optional<model::Resource>();
    if (stream.readBool()) {
        model::Resource produceResourceValue = readResource(stream);
        produceResource.emplace(produceResourceValue);
    }
    int produceAmount = stream.readInt();
    int produceScore = stream.readInt();
    bool harvest = stream.readBool();
    int workAmount = stream.readInt();
    return BuildingProperties(buildResources, maxHealth, maxWorkers, workResources, produceWorker, produceResource, produceAmount, produceScore, harvest, workAmount);
}

// Write BuildingProperties to output stream
void BuildingProperties::writeTo(OutputStream& stream) const {
    stream.write((int)(buildResources.size()));
    for (const auto& buildResourcesEntry : buildResources) {
        const model::Resource& buildResourcesKey = buildResourcesEntry.first;
        const int& buildResourcesValue = buildResourcesEntry.second;
        stream.write((int)(buildResourcesKey));
        stream.write(buildResourcesValue);
    }
    stream.write(maxHealth);
    stream.write(maxWorkers);
    stream.write((int)(workResources.size()));
    for (const auto& workResourcesEntry : workResources) {
        const model::Resource& workResourcesKey = workResourcesEntry.first;
        const int& workResourcesValue = workResourcesEntry.second;
        stream.write((int)(workResourcesKey));
        stream.write(workResourcesValue);
    }
    stream.write(produceWorker);
    if (produceResource) {
        stream.write(true);
        const model::Resource& produceResourceValue = *produceResource;
        stream.write((int)(produceResourceValue));
    } else {
        stream.write(false);
    }
    stream.write(produceAmount);
    stream.write(produceScore);
    stream.write(harvest);
    stream.write(workAmount);
}

// Get string representation of BuildingProperties
std::string BuildingProperties::toString() const {
    std::stringstream ss;
    ss << "BuildingProperties { ";
    ss << "buildResources: ";
    ss << "{ ";
    size_t buildResourcesIndex = 0;
    for (const auto& buildResourcesEntry : buildResources) {
        if (buildResourcesIndex != 0) {
            ss << ", ";
        }
        const model::Resource& buildResourcesKey = buildResourcesEntry.first;
        const int& buildResourcesValue = buildResourcesEntry.second;
        ss << resourceToString(buildResourcesKey);
        ss << ": ";
        ss << buildResourcesValue;
        buildResourcesIndex++;
    }
    ss << " }";
    ss << ", ";
    ss << "maxHealth: ";
    ss << maxHealth;
    ss << ", ";
    ss << "maxWorkers: ";
    ss << maxWorkers;
    ss << ", ";
    ss << "workResources: ";
    ss << "{ ";
    size_t workResourcesIndex = 0;
    for (const auto& workResourcesEntry : workResources) {
        if (workResourcesIndex != 0) {
            ss << ", ";
        }
        const model::Resource& workResourcesKey = workResourcesEntry.first;
        const int& workResourcesValue = workResourcesEntry.second;
        ss << resourceToString(workResourcesKey);
        ss << ": ";
        ss << workResourcesValue;
        workResourcesIndex++;
    }
    ss << " }";
    ss << ", ";
    ss << "produceWorker: ";
    ss << produceWorker;
    ss << ", ";
    ss << "produceResource: ";
    if (produceResource) {
        const model::Resource& produceResourceValue = *produceResource;
        ss << resourceToString(produceResourceValue);
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "produceAmount: ";
    ss << produceAmount;
    ss << ", ";
    ss << "produceScore: ";
    ss << produceScore;
    ss << ", ";
    ss << "harvest: ";
    ss << harvest;
    ss << ", ";
    ss << "workAmount: ";
    ss << workAmount;
    ss << " }";
    return ss.str();
}

}